meeting-notes/2020/Q2/2020-05-21--design-discussion-subdomains-dns-label-limit.md

```diff
@@ -0,0 +1,93 @@
+# Design Discussion: Subdomain support for CIDs longer than 63 characters
+
+## Context
+
+Full explanation: https://github.com/ipfs/go-ipfs/issues/7318
+
+**tl;dr**  
+
+[RFC 1034](https://tools.ietf.org/html/rfc1034#page-7) states that "each node has a label, 
+which is zero to 63 octets in length", this means CID in subdomain can't be longer than 63 characters.
+
+If we split at 63rd character that will produce TLS error because it is impossible 
+to have TLS cert for more than one wildcard level (`*.*.ipns.dweb.link`), as clarified in [RFC6125](https://tools.ietf.org/html/rfc6125#page-27)
+and implemented by browser vendors.
+
+DNS length limits were known, but we did not include inlined ed25519 in evaluation at a time, and this recent development changes the timeline, so we need to have a fixes for longer CIDs and TLS for IPNS websites.
+
+The Goal of this meeting is to find the least invasive path forward.
+
+
+## Participants
+
+- @Stebalien
+- @lidel
+- @ribasushi
+- @aschmahmann
+
+
+
+## Options vs Tradeoffs
+
+### Generic fix
+
+Split after 63rd char: "Do it anyway" ?
+add support, if not for ed25519, it will be useful for longer CIDs (they won't have TLS anyway) 
+
+
+### IPNS/ed25519-specific fixes
+
+Anything we can do to give IPNS websites created with ed25519 TLS without warnings?
+
+| Fix    | TLS for IPNS | TLS for long CID  | touches libp2p | different text representation in subdomain than base32 |
+| --- | --- | --- | --- | --- |
+| custom multicodec for ed25519 keys | YES | NO | YES | NO |
+| Allow PeerIDs as multihash in base32 (only in `*.ipns` subdomains) | YES | NO | NO | YES |
+| PeerIDs as CIDv1 in base36 | YES | NO | YES | YES |
+| "sha256 split in two" idea [missing notes, sorry!] | YES | ? | NO | YES |
+
+## Notes
+
+- all agree that touching key encoding in libp2p too expensive atm
+- leaking browser limitations down the stack is something we want to avoid, looking for a surgical fix that is compatible with CIDv1 specs
+
+- base36 makes ed25519 fit in a single DNS label
+  - missing from multibase - keeps need to register 
+  - could be used only for IPNS
+- AI: test 63 character domain on Cloudflare
+    - OK: confirmed 63 works: https://abc012345678901234567890123456789012345678901234567890123456789.lidel.org/ 
+      ("github pages" proxied via Cloudflare that gives automatic TLS cert)
+- **Plan for enabling TLS for IPNS with ed25519 on subdomain gateways**
+    - implement support for base36 in go and js (multibase support)
+    - ship with go-ipfs 0.6 ad js-ipfs, giving us one release overlap
+    -  in parallel implement splitting after 63rd char (for longer CIDs that are not ed25519 keys)    
```

Comment:
@lidel the splitting sub-spec probably could use a review: there are some details to be considered, e.g. is every individual sub-unit multibase-prefixed?

Is there a preexisting writeup/implementation somewhere I can look at and mull a bit?

Reply:
My plan was to open PR against go-ipfs and discuss it there. 

Initial  idea is to scope splitting to subdomain-gateways only and:

- when redirecting from `dweb.link/ipns/{key}` to → `{k.ey}.ipns.dweb.link` split at 63rd character to maximize the uniqueness of each label. 
- have normalization logic which redirects to "canonical split" version, to ensure people who manually put `.` in the middle of CID end up on the same Origin. 

I'll cc you on the PR when I get to it.